### PR TITLE
lexer: Disallow bare CR in raw byte strings

### LINKED ISSUE #60604: Raw Byte strings do not handle `\r`
Everywhere in the language we treat `\r\n` as `\n`, and forbid bare `\r`. In particular, in raw strings we do translate `\r\n` to `\n`, which makes sense, given that line ending depends on git config.

However, for raw byte string literals we don't do this, and this looks like a bug?

```
16:22:13|~/tmp
λ echo -e "fn main() {\n    let (s1, s2) = (br\"\r\", br\"\r\n\");\n    println!(\"{:?} {:?}\", s1, s2);\n}" > main.rs

16:25:49|~/tmp
λ rustc main.rs

16:25:50|~/tmp
λ ./main
[13] [13, 10]
```

### PR DESCRIPTION
**Pull request by Xanewok**
Handles bare CR ~but doesn't translate `\r\n` to `\n` yet in raw strings yet~ and translates CRLF to LF in raw strings.

As a side-note I think it'd be good to change the `unescape_` to return plain iterators to reduce some boilerplate (e.g. `has_error` could benefit from collecting `Result<T>` and aborting early on errors) but will do that separately, unless I missed something here that prevents it.

@matklad @petrochenkov thoughts?

### CODE REVIEW

File: src/libsyntax/lib.rs [hunk 1]
```diff
@@ -18,6 +18,7 @@
 #![feature(step_trait)]
 #![feature(try_trait)]
 #![feature(unicode_internals)]
+#![feature(rustc_private)]
```
Comment (Xanewok) [inline src/libsyntax/lib.rs]: The changes from the first commit (so `cargo check` works) were of drive-by nature, can remove them if needed

File: src/libsyntax/parse/unescape.rs [hunk 2]
```diff
@@ -66,6 +66,28 @@ where
...
+pub(crate) fn unescape_raw_str<F>(literal_text: &str, callback: &mut F)
+where
+    F: FnMut(Range<usize>, Result<char, EscapeError>),
+{
+    let mut byte_offset: usize = 0;
+
+    let mut chars = literal_text.chars().peekable();
+    while let Some(curr) = chars.next() {
+        let result = match (curr, chars.peek()) {
+            ('\r', Some('\n')) => Ok(curr),
```
Comment (matklad) [inline src/libsyntax/parse/unescape.rs]: Looks like some special handling is missing here:

* return `\n`
*  bump `chars` one more time
* make byte_offset.. range spanning two chars

File: src/libsyntax/parse/unescape.rs [hunk 3]
```diff
@@ -88,6 +89,29 @@ where
...
+pub(crate) fn unescape_raw_byte_str<F>(literal_text: &str, callback: &mut F)
+where
+    F: FnMut(Range<usize>, Result<char, EscapeError>),
+{
+    let mut byte_offset: usize = 0;
+
+    let mut chars = literal_text.chars().peekable();
+    while let Some(curr) = chars.next() {
+        let result = match (curr, chars.peek()) {
+            ('\r', Some('\n')) => Ok(curr),
```
Comment (matklad) [inline src/libsyntax/parse/unescape.rs]: Same here, should be `\n` and not `curr`, whihc is `\r`

File: src/test/ui/parser/raw-byte-string-literals.rs [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 
 pub fn main() {
+    br"a"; //~ ERROR bare CR not allowed in string
     br"é";  //~ ERROR raw byte string must be ASCII
     br##~"a"~##;  //~ ERROR only `#` is allowed in raw string delimitation
 }
```
Comment (matklad) [inline src/test/ui/parser/raw-byte-string-literals.rs]: Let's also add a test that \r\n is translated to \n?

I guess a run-pass `assert_eq!(rb"\r\n", [10])` should do the trick

File: src/libsyntax/parse/unescape.rs (hunk 2 above)
Comment (matklad) [reply src/libsyntax/parse/unescape.rs]: Ah, the description of PR says that this is a WIP :)

**Comment by matklad** [comment]
>As a side-note I think it'd be good to change the unescape_ to return plain iterators to reduce some boilerplate (e.g. has_error could benefit from collecting Result<T> and aborting early on errors) but will do that separately, unless I missed something here that prevents it.

I *think* changing `unescape_` to usual, external iterators, will make implementation trickier and slower in the common case: one would have to reify the state, instead of keeping it implicitly on the stack. If I am mistaking and the impl turns out not significantly worse, I am all for flipping to external iterators!

However, I think here it's counter-intiutively more important to preserve implementation simplicity (that's where all interesting bits are) rather than interface simplicity (interface is a glue code, which is isolated), so if internal iteration is easier to implement, I'd keep it.

File: src/libsyntax/parse/literal.rs [hunk 5]
```diff
@@ -4,7 +4,8 @@ use crate::ast::{self, Ident, Lit, LitKind};
 use crate::parse::parser::Parser;
 use crate::parse::PResult;
 use crate::parse::token::{self, Token};
-use crate::parse::unescape::{unescape_str, unescape_char, unescape_byte_str, unescape_byte};
+use crate::parse::unescape::{unescape_str, unescape_char, unescape_byte_str,
+    unescape_raw_str, unescape_byte};
```
Comment (petrochenkov) [inline src/libsyntax/parse/literal.rs]: Nit: could you keep the "one line - one `use`" formatting.

File: src/libsyntax/lib.rs (hunk 1 above)
Comment (petrochenkov) [reply src/libsyntax/lib.rs]: Yes, please remove.

**Comment by petrochenkov** [comment]
>Closes #60604.

I'd keep https://github.com/rust-lang/rust/issues/60604 open until the `\r\n` -> `\n` translation is moved to the file loading stage (or a decision that we don't want to move it is made).

**Comment by petrochenkov** [comment]
The direction seems right (same as https://github.com/rust-lang/rust/pull/60261 basically).
I'll wait with review until this fully ready.

File: src/libsyntax/lib.rs (hunk 1 above)
Comment (Xanewok) [reply src/libsyntax/lib.rs]: Done. Out of curiosity, why should  we skip this? Is it because it adds noise and is somewhat obvious? (after all we're in *the* compiler code)

File: src/libsyntax/parse/unescape.rs (hunk 2 above)
Comment (Xanewok) [reply src/libsyntax/parse/unescape.rs]: If possible, I'd like to do this as a separate PR :)

File: src/libsyntax/parse/literal.rs (hunk 5 above)
Comment (Xanewok) [reply src/libsyntax/parse/literal.rs]: Done

**Comment by Xanewok** [comment]
> (...) will make implementation trickier and slower in the common case: one would have to reify the state, instead of keeping it implicitly on the stack

@matklad maybe we could use [`iter::from_fn`](https://doc.rust-lang.org/std/iter/fn.from_fn.html)?

Rebased the PR and would like for this to land and then land a separate one which translates \r\n -> \n if possible.

**Comment by Xanewok** [comment]
```
[01:05:24] failures:
[01:05:24] 
[01:05:24] ---- [run-pass] run-pass/lexer-crlf-line-endings-string-literal-doc-comment.rs stdout ----
[01:05:24] 
[01:05:24] error: test run failed!
[01:05:24] status: exit code: 101
[01:05:24] thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:521:22
[01:05:24] note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
[01:05:24] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/lexer-crlf-line-endings-string-literal-doc-comment/a"
[01:05:24] stdout:
[01:05:24] ------------------------------------------
[01:05:24] 
[01:05:24] ------------------------------------------
[01:05:24] stderr:
[01:05:24] ------------------------------------------
[01:05:24] thread 'main' panicked at 'assertion failed: `(left == right)`
[01:05:24]   left: `"string\r\nliteral"`,
[01:05:24]  right: `"string\nliteral"`', /checkout/src/test/run-pass/lexer-crlf-line-endings-string-literal-doc-comment.rs:32:5
[01:05:24] note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
[01:05:24] 
```
My bad, it seems that CRLF -> LF  transcription should be done as well in this PR with the validation logic moved

File: src/test/ui/parser/raw-byte-string-literals.rs (hunk 4 above)
Comment (Xanewok) [reply src/test/ui/parser/raw-byte-string-literals.rs]: As it turns out we already have a test for that:
https://github.com/rust-lang/rust/blob/5c6f6b810c1d5850814a3c5960c4dfe2aee77d0d/src/test/run-pass/lexer-crlf-line-endings-string-literal-doc-comment.rs#L30-L32

File: src/libsyntax/parse/unescape.rs (hunk 3 above)
Comment (Xanewok) [reply src/libsyntax/parse/unescape.rs]: Done in https://github.com/rust-lang/rust/pull/60793/commits/d4632744fa0fb18a7c3f5058f1e8157c760353b4

File: src/libsyntax/parse/unescape.rs (hunk 2 above)
Comment (Xanewok) [reply src/libsyntax/parse/unescape.rs]: (hopefully correctly) done in https://github.com/rust-lang/rust/pull/60793/commits/d4632744fa0fb18a7c3f5058f1e8157c760353b4

**Comment by Xanewok** [comment]
Pushed the CRLF -> LF translation now and tests seem to pass locally now, hopefully CI agrees :crossed_fingers:

File: src/libsyntax/parse/lexer/mod.rs [hunk 6]
```diff
@@ -1315,16 +1242,16 @@ impl<'a> StringReader<'a> {
         id
     }
 
-    fn scan_raw_byte_string(&mut self) -> (token::LitKind, Symbol) {
+    fn scan_raw_string(&mut self) -> (BytePos, BytePos, u16) {
```
Comment (matklad) [inline src/libsyntax/parse/lexer/mod.rs]: ` Let's add a comment explaining what `(BytePos, BytePos, u16)` mean?

File: src/libsyntax/parse/unescape.rs [hunk 7]
```diff
@@ -66,6 +67,35 @@ where
...
+                chars.next();
+                (Ok('\n'), [Some('\r'), Some('\n')])
+            },
+            ('\r', _) =>
+                (Err(EscapeError::BareCarriageReturn), [Some('\r'), None]),
+            (c, _) if mode.is_bytes() && c > '\x7F' =>
+                (Err(EscapeError::NonAsciiCharInByteString), [Some(c), None]),
+            (c, _) => (Ok(c), [Some(c), None]),
+        };
+        let len_utf8: usize = scanned.iter().filter_map(|&x| x).map(char::len_utf8).sum();
```
Comment (matklad) [inline src/libsyntax/parse/unescape.rs]: I think we don't need to maintain `scanned` separately, because it duplicates position info that is already present in the chars iterator. I suggest restoring this info from `chars.as_str().len()`

File: src/test/ui/parser/raw-byte-string-literals.rs (hunk 4 above)
Comment (matklad) [reply src/test/ui/parser/raw-byte-string-literals.rs]: That tests raw strings, and not raw *byte* strings. Let's extend that test to raw byte strings as well: after all, this is the bug we want to fix with this change :D

Comment (Xanewok) [reply src/test/ui/parser/raw-byte-string-literals.rs]: Right :man_facepalming: Adding a test now

File: src/libsyntax/parse/lexer/mod.rs (hunk 6 above)
Comment (Xanewok) [reply src/libsyntax/parse/lexer/mod.rs]: Will do!

File: src/libsyntax/parse/unescape.rs (hunk 7 above)
Comment (Xanewok) [reply src/libsyntax/parse/unescape.rs]: I like how this version is more declarative (I find `initial_len`, `start`, `end` calculation dance [here](https://github.com/rust-lang/rust/blob/07c3967de9027735bfbf9c739f6a8ca14312c22d/src/libsyntax/parse/unescape.rs#L211-L243) not immediately obvious) but then you have to manually duplicate what you just scanned, which isn't ideal.

I can rewrite it using the start/end calculation if you'd like

Comment (matklad) [reply src/libsyntax/parse/unescape.rs]: I slightly prefer to rewrite: I agree that it is spectacularly non-obvious, but it is guaranteed to be correct. in contrast, with the current approach one may accidentally bump twice, but return only one char in scanned.

Comment (Xanewok) [reply src/libsyntax/parse/unescape.rs]: Done in https://github.com/rust-lang/rust/pull/60793/commits/63dc7da703759dd53536dd18a42ff65f39a2f9b4 - I think it's better this way, thanks!

**Comment by Xanewok** [comment]
Applied feedback and actually implemented CRLF -> LF rewrite in raw byte strings (forgot about the byte variant :man_facepalming: ) in https://github.com/rust-lang/rust/pull/60793/commits/735ac057bb94d179c81afcd0f3e63bc6e6856734#diff-a8037ce27e05cd3a29b041f9ec00f94eR176

**Comment by matklad** [comment]
LGTM now, let's wait for @petrochenkov review!

File: src/test/ui/parser/raw-byte-string-literals.stderr [hunk 8]
```diff
@@ -1,14 +1,20 @@
-error: raw byte string must be ASCII: \u{e9}
-  --> $DIR/raw-byte-string-literals.rs:2:8
+error: bare CR not allowed in string, use \r instead
```
Comment (petrochenkov) [inline src/test/ui/parser/raw-byte-string-literals.stderr]: The "use \r instead" part is not applicable to raw strings.

**Comment by petrochenkov** [comment]
r=me with https://github.com/rust-lang/rust/pull/60793#discussion_r292024803 fixed

File: src/libsyntax/lib.rs (hunk 1 above)
Comment (petrochenkov) [reply src/libsyntax/lib.rs]: Less noise and impossible to forget something if it's added automatically.
rustc crates are not supposed to be built in isolation, outside of the rustc build system, so boilerplate common for all rustc crates is regularly moved to the workspace or rustbuild.
The `rustc_private` features in particular were moved in https://github.com/rust-lang/rust/pull/41847.

Comment (petrochenkov) [reply src/libsyntax/lib.rs]: IIRC, `x.py check` should provide the same functionality as `cargo check`.

File: src/test/ui/parser/raw-byte-string-literals.stderr (hunk 8 above)
Comment (Xanewok) [reply src/test/ui/parser/raw-byte-string-literals.stderr]: Good catch, fixed!
